src/dma/mod.rs

```diff
@@ -465,115 +448,128 @@ where
-
-                // double buffering, unwrap can never fail
-                return Ok((old_buf.unwrap(), CurrentBuffer::DoubleBuffer));
+                (true, inactive)
             }
-        }
-        self.stream.disable();
-        self.stream.clear_transfer_complete_interrupt();
+        };
 
-        // "No re-ordering of reads and writes across this point is allowed"
+        // Protect the instruction sequence of preceding DMA disable/inactivity
+        // verification/poisoning and subsequent (old completed) buffer content
+        // access.
         compiler_fence(Ordering::SeqCst);
+        // Also protect the corresponding data access sequence.
+        // NOTE: The data cache also needs to be flushed (if used).
+        cortex_m::asm::dmb();
 
-        // NOTE(unsafe) We now own this buffer and we won't call any &mut
-        // methods on it until the end of the DMA transfer
-        let buf_len = unsafe {
-            let (buf_ptr, buf_len) = new_buf.write_buffer();
+        // This buffer is inactive now and can be accessed.
+        // NOTE(panic): We always hold ownership in lieu of the DMA peripheral.
+        let buf = self.buf[inactive as usize].take().unwrap();
 
-            self.stream.set_memory_address(buf_ptr as u32);
-            buf_len
-        };
-        self.stream.set_number_of_transfers(buf_len as u16);
-        let old_buf = self.buf.replace(new_buf);
+        let (mut buf, result) = func(buf, inactive);
 
-        // Ensure that all transfers to normal memory complete before
-        // subsequent memory transfers.
-        //
-        // The new memory buffer is almost certainly in normal memory, so we
-        // ensure that all transfers there complete before proceeding to enable
-        // the stream
+        // NOTE(unsafe) We now own this buffer and we won't access it
+        // until the end of the DMA transfer.
+        let (buf_ptr, buf_len) = unsafe { buf.write_buffer() };
+
+        // Keep ownership of the active buffer in lieu of the DMA peripheral.
+        self.buf[inactive as usize].replace(buf);
+
+        // Protect the instruction sequence of preceding (new) buffer content access
+        // and subsequent DMA enable/address update.
+        compiler_fence(Ordering::SeqCst);
+        // Also protect the corresponding data access sequence.
+        // NOTE: The data cache also needs to be flushed (if used).
         cortex_m::asm::dmb();
 
-        // "Preceding reads and writes cannot be moved past subsequent writes"
-        compiler_fence(Ordering::Release);
+        if !double {
+            self.stream.set_number_of_transfers(buf_len as u16);
+        } else if buf_len != usize::from(self.transfer_length) {
+            // We can't change the transfer length while double buffering
+            return Err(DMAError::SmallBuffer);
+        }
 
         unsafe {
-            self.stream.enable();
+            self.stream.set_memory_address(inactive, buf_ptr as usize);
```

Comment:
Would you mind adding a comment here that we can ignore failures to write this address in double buffer mode specifically because of the address poisoning that we invoked earlier? That is, _if_ this buffer has already become active and a write would fail, the DMA has gone into an error state and the stream has disabled anyways, so this write is guaranteed to pass.

Adding this note will likely help us in the future determine why this is acceptable (and why it may not be if we refactor)

Reply:
It's not like the write failing or passing in that case matters. In fact it always passes from the CPU perspective.